src/axom/quest/MarchingCubes.hpp

```diff
@@ -0,0 +1,295 @@
+#include "axom/config.hpp"
+#include "axom/mint/mesh/Mesh.hpp"
+
+// Conduit includes
+#include "conduit_node.hpp"
+
+// C++ includes
+#include <string>
+
+namespace axom
+{
+namespace quest
+{
+/*!
+ * \@brief Class implementing marching cubes algorithm on a single
+ * structured mesh domain within a multi-domain mesh.
+ *
+ * Implementation is for 2D (marching squares) and 3D (marching
+ * cubes).
+ *
+ * The input mesh is a Conduit::Node following the Mesh Blueprint
+ * convention.
+ *
+ * Usage example:
+ * @beginverbatim
+ *   MarchingCubes computationalMesh(meshNode, "coords");
+ *   axom::mint::UnstructuredMesh<axom::mint::_SINGLe_SHAPE>
+ *     surfaceMesh(3, min::CellType::Triangle);
+ *   double contourValue = 0.0;
+ *   set_function_field("my_function");
+ *   set_output_mesh(surfaceMesh);
+ *   mc.compute_iso_surface(contourValue);
+ * @endverbatim
+ *
+ * \sa MarchingCubes
+ */
+class MarchingCubesSingleDomain
+{
+public:
+  /*!
+   * \brief Constructor for applying algorithm in multi-level mesh context.
+   *
+   * \param [in] dom Blueprint single-domain mesh containing scalar field.
+   * \param [in] coordsetName Name of blueprint point coordinate set.
+   * \param [in] maskField Cell-based std::int32_t mask field.  If provided,
+   *             cells where this field evaluates to false are skipped.
+   *
+   * Some data from \a dom may be cached by the constructor.
+   * Any change to it after the constructor leads to undefined behavior.
+   * See set_domain(const conduit::Node &) for requirements on \a dom.
+   *
+   * The mesh coordinates should be contiguous.  See
+   * conduit::blueprint::is_contiguous().  In the future, this
+   * requirement may be relaxed, possibly at the cost of a
+   * transformation and storage of the temporary contiguous layout.
+   */
+  MarchingCubesSingleDomain(const conduit::Node &dom,
+                            const std::string &coordsetName,
+                            const std::string &maskfield);
+
+  //! @brief Spatial dimension of domain.
+  int dimension() const { return m_ndim; }
+
+  /*!
+    @brief Set the field containing the nodal function.
+    \param [in] fcnField Name of node-based scalar function values.
+  */
+  void set_function_field(const std::string &fcnField);
+
+  /*!
+    @brief Set the output surface mesh object.
+  */
+  void set_output_mesh(axom::mint::Mesh *surfaceMesh);
+
+  /*!
+    @brief Save the originating cell index in output mesh.
+
+    For each contour mesh cell generated, the 1D index of the
+    generating computational mesh cell is set in the \a cellIdField of
+    the contour mesh.  If the field doesn't exist, it will be created.
+    To disable, set \a outputCellIds to empty.
+  */
+  void set_cell_id_field(const std::string &cellIdField)
+  {
+    m_cellIdField = cellIdField;
+  }
+
+  /*!
+   * \brief Computes the iso-surface.
+   *
+   * \param [in] contourVal iso-contour value
+   *
+   * Compute iso-surface using the marching cubes algorithm.
```

Comment:
Please discuss what required object state previous to calling `compute_iso_surface()`.  For example, "A code must call set_function_field() and set_output_mesh() before calling compute_iso_surface.  Otherwise, an error will occur."  Or rephrase that however seems best.

This should relate to the usage example for MarchingCubes class.